[LAA][LV]Allow recognition of strided pointers with constant stride

nasherm opened the pull request:
This patch fixes an issue found during LoopAccessAnalysis with respect
to recognizing strided pointers that make use of runtime constants. Loop 
accesses of the form `p[base + offset * const]` , where `const` is a runtime constant
should be considered for vectorization. However, it was found that there 
were cases that these access patterns weren't recognized. This patch resolves
this by adding an explicit pattern match within LAA.

nasherm committed:
[IndVarSimplify] Add rewriting ptr-add phis with offset addressing

This patch adds support to IndVarSimplify to rewrite IVs used in loops for
strided loads or stores. These IVs are rewritten to make use of offset
addressing thereby elimnating PHI nodes and simplifying the IR.
The motivation for this is that with these PHI nodes eliminated vectorization
becomes more feasible.

Change-Id: Ib3385ee5de03c46b4de035b550fbf56403e9e376

nasherm committed:
Compiler warning due to use of deprecated function

Change-Id: Id8287887c3c77c24619ebf07b5f76a532217aff6

nasherm committed:
Formatting

Change-Id: I4d161c236398dca05be84d43e33ada913309b992

nasherm committed:
Copilot review

Change-Id: Ida6175d40f7bc0f95b28e37a30538aa3cbd64a1e

nasherm committed:
[LAA][LV]Allow recognition of strided pointers with constant stride

This patch fixes an issue found during LoopAccessAnalysis with respect
to recognising strided pointers that make use of runtime constants.
It was found that for a loop of the following form

```
for (unsigned row = 0; row < nRows; row++) {
  float *px = pOut + row;
  for (unsigned col = 0; col < nCols; col++) {
    *px = *pIn++;
    px += nRows;
}
```

Vectorization would expectedly fail due to the use of a runtime strided IV px.
When passing the `lv-strided-pointer-ivs=true` option, vectorization would be
attempted but would be seen as non-profitable. However, when transforming the loop
to the following form

```
for (unsigned row = 0; row < nRows; row++) {
  for (unsigned col = 0; col < nCols; col++) {
    pOut[row + col * nRows] = *pIn++;
}
```

With `lv-strided-pointer-ivs=true` this vectorizes successfully. As these
loops are equivalent, investigation sought to find out why vectorization was failing.
The cause was found to be that when obtaining the stride for the IVs LAA doesn't
recognise strides that may make use of loop-invariant runtime constants
unless they are in a specific SCEV add-rec form such as in loop two. This patch
allows for the first type of add-rec to also be recognised when the option
`lv-strided-pointer-ivs=true` is passed to the compiler.

Change-Id: Ib01a6b802f516516ac0215569a8d283af915c297

david-arm commented on llvm/test/Transforms/LoopVectorize/RISCV/strided-accesses.ll:
```diff
@@ -861,33 +856,24 @@ define void @double_stride_int_scaled(ptr %p, ptr %p2, i64 %stride) {
-; STRIDED-UF2-NEXT:    call void @llvm.masked.scatter.nxv4i32.nxv4p0(<vscale x 4 x i32> [[TMP38]], <vscale x 4 x ptr> align 4 [[TMP40]], <vscale x 4 x i1> splat (i1 true)), !alias.scope [[META11]], !noalias [[META8]]
+; STRIDED-UF2-NEXT:    [[TMP36:%.*]] = getelementptr i32, ptr [[P2]], i64 [[TMP31]]
+; STRIDED-UF2-NEXT:    [[TMP39:%.*]] = getelementptr i32, ptr [[TMP36]], i64 [[TMP29]]
+; STRIDED-UF2-NEXT:    store <vscale x 4 x i32> [[TMP37]], ptr [[TMP36]], align 4, !alias.scope [[META11:![0-9]+]], !noalias [[META8]]
```
Perhaps I'm missing something, but this doesn't look right to me. The pattern of loads is:

Load 1: load i32, %p
Load 2: load i32, %p + %stride
Load 3: load i32, %p + (2 * %stride)
Load 4: load i32, %p + (3 * %stride)

The total range of memory accessed is `[%p, %p + (3 * %stride) + 4)`. A v4i32 normal vector load would access memory in the range `[%p, %p + 16)`. This can only be valid if `%stride == 1`, which would require SCEV checks before entering the vector loop. Essentially, there are holes between the loads/stores, which I think is why we previously used gathers and scatters.

fhahn commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -896,7 +896,8 @@ getPtrStride(PredicatedScalarEvolution &PSE, Type *AccessTy, Value *Ptr,
              const Loop *Lp, const DominatorTree &DT,
```
Currently callers of the function currently expect this to return the exact stride, but with the new flag it currently seems to ignore the multiplier. Passing `AllowStridedPtrs =true` will break the expectation of the current users AFAICT, causing incorrect results.

nasherm committed:
Add const runtime strided pointer helpers

Change-Id: I554dba37cb16436851427d47298733784d08ff6c

nasherm committed:
Plugging use of runtime const stride helpers into vectorizer

- makes use of new LAA methods
- adds changes to LV cost model decision code to allow for widening
  or gather/scatter.

Change-Id: I108ad4a995d54f7f40b84ba8abdc362cc6681125

nasherm committed:
Review comments

- refactor runtime strided pointer checks to fit
  into the setCostBasedWideningDecision->isConsecutivePtr
  ->getPtrStride call pipeline.
- refactoring comments, updating tests.

Change-Id: I5f16032c8841edc861c4779999efae086ff91469

nasherm commented:
Ping

nasherm committed:
Review comments

- make use of gather scatter cost for cost widening of runtime strided ptrs
- re-introduced isConstRuntimeStridedPtr method
- update tests

Change-Id: Ie9dd939c871488cb65664e4196a2cfac4e0eaf53

fhahn reviewed (commented):
I am not sure if we need all that complexity.

We already have support for versioning strides, it just looks like multiplications with constants are not supported yet.

Would it be enough to just extend `getStrideFromPointer` to support that stride pattern?

```
 diff --git a/llvm/lib/Analysis/LoopAccessAnalysis.cpp b/llvm/lib/Analysis/LoopAccessAnalysis.cpp
index 2fc724970747..8a2c6880bc68 100644
--- a/llvm/lib/Analysis/LoopAccessAnalysis.cpp
+++ b/llvm/lib/Analysis/LoopAccessAnalysis.cpp
@@ -2979,6 +2979,8 @@ static const SCEV *getStrideFromPointer(Value *Ptr, ScalarEvolution *SE, Loop *L
   if (isa<SCEVUnknown>(V))
     return V;

+  const SCEVUnknown *UK;
+  match(V, m_scev_Mul(m_SCEVConstant(), m_SCEV(V)));
   if (auto *C = dyn_cast<SCEVIntegralCastExpr>(V))
     if (isa<SCEVUnknown>(C->getOperand()))
       return V;
```?

nasherm committed:
Update getPtrStride to recognise const runtime strided patterns

Update the getPtrStride method to allow recognition of runtime
const strided pointers.

Change-Id: I67b41c0f0948542b2f07d6d3fea1cee826a4d0cf

nasherm committed:
Comment typos

Change-Id: Id3f796bfb68b2b8b8c1bd5f7c053aff2fdcade9e

nasherm committed:
Add induction test

Change-Id: I04a289ca7cc83753b69a7c645ba5a426159f3b94

nasherm commented on llvm/test/Transforms/LoopVectorize/pointer-induction.ll:
```diff
@@ -20,65 +20,65 @@ define void @a(ptr readnone %b) {
 ; CHECK-NEXT:    [[N_VEC:%.*]] = sub i64 [[TMP0]], [[N_MOD_VF]]
 ; CHECK-NEXT:    [[TMP1:%.*]] = mul i64 [[N_VEC]], -1
-; CHECK-NEXT:    [[IND_END:%.*]] = getelementptr i8, ptr null, i64 [[TMP1]]
+; CHECK-NEXT:    [[TMP2:%.*]] = getelementptr i8, ptr null, i64 [[TMP1]]
```
Removed

nasherm commented on llvm/test/Transforms/LoopVectorize/RISCV/strided-accesses.ll:
```diff
@@ -329,15 +329,15 @@ define void @single_stride_int_scaled(ptr %p, i64 %stride) {
+; NOSTRIDED-NEXT:    [[WIDE_MASKED_GATHER:%.*]] = call <vscale x 4 x i32> @llvm.vp.load.nxv4i32.p0(ptr align 4 [[TMP1]], <vscale x 4 x i1> splat (i1 true), i32 [[TMP7]])
+; NOSTRIDED-NEXT:    [[TMP4:%.*]] = add <vscale x 4 x i32> [[WIDE_MASKED_GATHER]], splat (i32 1)
+; NOSTRIDED-NEXT:    call void @llvm.vp.store.nxv4i32.p0(<vscale x 4 x i32> [[TMP4]], ptr align 4 [[TMP1]], <vscale x 4 x i1> splat (i1 true), i32 [[TMP7]])
 ; NOSTRIDED-NEXT:    [[TMP11:%.*]] = zext i32 [[TMP7]] to i64
```
Removed

nasherm committed:
Merge branch 'main' into nashe/mat-transpose-fail-autovec

fhahn reviewed (approved):
LGTM, thanks

Would be good to update the commit message to clarify the change and why it is needed. As is the description is quite verbose and could do with some tightening, focusing on the rationale behind the latest version of diff

nasherm committed:
Update llvm/lib/Analysis/LoopAccessAnalysis.cpp

Co-authored-by: Florian Hahn <flo@fhahn.com>

nasherm committed:
Code formatting

Change-Id: I6802f443651b7e9e9eaef855d096ef2dce0c0665

eas commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2975,6 +2975,8 @@ static const SCEV *getStrideFromPointer(Value *Ptr, ScalarEvolution *SE, Loop *L
+  match(V, m_scev_Mul(m_SCEVConstant(), m_SCEV(V)));
   if (auto *C = dyn_cast<SCEVIntegralCastExpr>(V))
     if (isa<SCEVUnknown>(C->getOperand()))
       return V;
```
I'm getting familiar with the LAA code by studying recent patches, so forgive me if the question is stupid. How is this correct to return `V` as stride if we stripped multiplication by a constant above?

In fact, pre-existing looking through the GEP itself is wrong (IMO) - see https://github.com/llvm/llvm-project/pull/182595#discussion_r2854669701.

Are you trying to speculate `V` to be one because you want some memory dependences checks to be emitted at all and not because you want some access to become unit-striped?

eas replied:
> **The result should only be used** to add a SCEV assumption %stride == 1, and the full pointer SCEV we use contructed from the original pointer via PSE.getSCEV(Ptr)

[Emphasis mine] Doesn't that mean that current function name is misleading and its documentation is insufficient at best?

> In cases where the constant multiplier matches the access size

It seems some other code verifies (I couldn't make https://godbolt.org/z/bbnhb3cv3 produce "bad" code and I don't see such a check anywhere around new code) that that match happens indeed, but isn't that extremely confusing that those two parts of the condition are checked so far apart?
